Use Github Action to replace Travis and AppVeyor for CI.

wyk9787 opened the pull request:
## TL;DR

### Things included in this PR:

* Added a Github Action script **draft** (macOS, Linux and Windows test).

### Follow-up plan:

* Verify the correctness of the draft.

* Remove working {Go, OS, XCode} sections that are covered by this Github Action script from existing CI scripts.

* Add build badge to README.md.

* Enable Github Action as part of PR review process.

## Details

### Why it is only a draft?

The Github Action script (i.e. ci.yaml) is first introduced in this PR. That means any change made to ci.yaml can be only tested on the forked repo (i.e. it will trigger the Github Action on the forked repo if a commit is pushed to the master branch). This seems to be the intentional behavior as noted [here](https://github.community/t/run-a-github-action-on-pull-request-for-pr-opened-from-a-forked-repo/16054).

### Limitations

1. The master branch of Go is not supported by setup-go. There is an open PR that is working on this: https://github.com/actions/setup-go/pull/34, but no progress has been made since March and no ETA.

2. Only the latest macOS is supported: https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idruns-on.

3. Limited number of XCode versions are supported: https://github.com/actions/virtual-environments/blob/master/images/macos/macos-10.15-Readme.md#xcode, but this should not be an issue for us as we wouldn't support anything older than XCode 10.3.

#### Solution

We will keep the tests for {GO, OS, XCode} combination that are **not** currently supported by Github Action in existing CI scripts. Ideally, we will only need to maintain one CI.

aalexand commented:
The series of commits with "Update to test" message is not very useful. Either make commit messages useful or squash the commits and "push -f".

aalexand commented:
The coverage report seems to say that the coverage has dropped - why?

wyk9787 commented:
> The series of commits with "Update to test" message is not very useful. Either make commit messages useful or squash the commits and "push -f".

This PR is still work in progress; there is "DO NO REVIEW" in PR description. I thought I needed to open the PR to test something. I will squash those commits once it is ready for review.

aalexand commented:
The Travis CI build is marked as red. Should we disable this? Also, can we mark the new CI configuration as required to pass for any PR to be submitted? We want to ensure that only green testing status is allowed to be submitted.

aalexand commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,99 @@
+          go env
+          go build github.com/google/pprof
+          go test -v ./...
+
```
nit: remove trailing empty lines

wyk9787 replied:
Done.

aalexand commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,99 @@
+    runs-on: windows-latest
+    strategy:
+      matrix:
+        go: ['1.13', '1.14']
```
Is it possible to list the supported Go versions only once in this file?

wyk9787 replied:
Unfortunately, it seems like the only variable it supports is environment variable, which does not work for list. https://help.github.com/en/actions/configuring-and-managing-workflows/using-variables-and-secrets-in-a-workflow

aalexand replied:
Can we use YAML references? https://blog.daemonl.com/2016/02/yaml.html

wyk9787 replied:
Not sure if you can see the error message here: https://github.com/wyk9787/pprof/actions/runs/124158614, but it's not allowed. I think Github Action has its own parser to make sure it respects Github Action's syntax, even though it is in the format of a yaml file: https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions.

aalexand replied:
Ok, looks like we'll have to leave in the duplication if you say there is no way to avoid it.

wyk9787 commented:
> The Travis CI build is marked as red. Should we disable this? Also, can we mark the new CI configuration as required to pass for any PR to be submitted? We want to ensure that only green testing status is allowed to be submitted.

The Travis CI still suffers from the timeout issue. We should disable this and remove travis.yml once we verify Github action is working correctly. As far as this PR, we should just keep retrying the Travis test to make it pass. Starting from next PR, which I will create once this one gets merged, we can get rid of Travis and start using Github Action.

Once this PR gets merged, I think admin (I don't have that access) is able to mark the new CI configuration as required to pass for any PR to be submitted in this repo's settings.

kalyanac commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,95 @@
+    strategy:
+      matrix:
+        go: ['1.13', '1.14']
+        os: ['ubuntu-latest', 'macos-latest']
```
I think Travis CI uses 2 versions on MacOS, Ubuntu 16.04 LTS. 
We should run on Ubuntu latest, and LTS at least. Similarly, latest 2 versions on macOS with latest 2 versions on Xcode.

wyk9787 replied:
Added 18.04 and 20.04. "ubuntu-latest" is 18.04. There is no support for other versions of macOS besides the latest, unfortunately. See [here](https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#supported-runners-and-hardware-resources).

I have spent the past two hours trying to mess around the keywords "include" and "exclude" as documented [here](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-including-additional-values-into-combinations). I am not sure this is a bug or it's the intended behavior, but basically there is no elegant way for me to say I want the additional following two combination: os = 'macos-latest', xcode-version = ['1.13', '1.14']. 

It's pretty clear from this [example](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-including-additional-values-into-combinations) that you are able to introduce a new variable here. But first of all you are not able to say:
```
include:
  - os: 'macos-latest'
     xcode-version: ['latest', '10.3']
```
because of the syntax issue (i.e. it does not support a list here).

But then you are not able to say
```
include:
  - os: 'macos-latest'
     xcode-version: 'latest'
  - os: 'macos-latest'
     xcode-version: '10.3'
```
either, since the second one (i.e. 10.3) will simply override the first one.

The final outcome is a compromise where I don't use "include" or "exclude" keywords at all and simply let all of them run. This will result 12 runs (see [here](https://github.com/wyk9787/pprof/runs/736723446?check_suite_focus=true)), where 4 of them are repeated (i.e. 2 ubuntu versions * 2 xcode versions). They will still pass without issues but those 4 runs are kind of wasted.

Given the time constraints on Github Action is really not an issue (with 12 runs, they finished in less than 5 minutes), I think this is the best we can do. I do notice Github Action is rapidly developing (e.g. this whole matrix.strategy thing is introduced not too long ago), so maybe there is hope afterall.

nolanmar511 replied:
If I recall correctly, there are either limits on the number of GitHub actions workflows or they're paid; also, more tests creates more opportunity for flakiness, and the listing of tests run might be somewhat confusing. 

It might be worth some repetition (specifically, splitting out Mac and Linux test cases) to avoid this.

kalyanac commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,95 @@
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        go: ['1.13', '1.14']
```
Include master too.

wyk9787 replied:
Master is not currently supported. See the last issue in PR description.

aalexand commented:
I'd be unhappy to lose testing against Go master branch.

wyk9787 commented:
> I'd be unhappy to lose testing against Go master branch.

One compromise we can make while we are waiting for https://github.com/actions/setup-go/pull/34 to happen is to keep the Travis CI but only testing against Go master branch. Supposedly, timeout should not be an issue in this case if we remove other Go versions.

aalexand commented:
> > I'd be unhappy to lose testing against Go master branch.
> 
> One compromise we can make while we are waiting for [actions/setup-go#34](https://github.com/actions/setup-go/pull/34) to happen is to keep the Travis CI but only testing against Go master branch. Supposedly, timeout should not be an issue in this case if we remove other Go versions.

There isn't recent progress on https://github.com/actions/setup-go/pull/34 it seems. We should probably ping it to try to get an estimate on whether it's coming and when.

wyk9787 commented:
> > > I'd be unhappy to lose testing against Go master branch.
> > 
> > 
> > One compromise we can make while we are waiting for [actions/setup-go#34](https://github.com/actions/setup-go/pull/34) to happen is to keep the Travis CI but only testing against Go master branch. Supposedly, timeout should not be an issue in this case if we remove other Go versions.
> 
> There isn't recent progress on [actions/setup-go#34](https://github.com/actions/setup-go/pull/34) it seems. We should probably ping it to try to get an estimate on whether it's coming and when.

Done.

wyk9787 commented:
https://github.com/actions/setup-go/pull/34 has not had any updates since end of March and did not reply to my comments either. Should we move forward with this PR? 

Again, this PR will NOT replace the Travis CI and AppVeyor yet. The next PR should be able to verify whether issues 1 and 2 in PR description is still a problem and then replace Travis CI and AppVeyor.

aalexand commented:
I'll leave it to @kalyanac and @nolanmar511 to decide.

nolanmar511 commented:
The Travis timeouts are pretty annoying. 

Perhaps we migrate windows tests (and get rid of appveyor), and migrate the MacOS tests that we can migrate. 

This will keep the number of distinct systems for CIs we use at 2 and should help reduce MacOS timeouts we experience on Travis. 

@kalyanac -- What are your thoughts?

aalexand commented:
> The Travis timeouts are pretty annoying.
> 
> Perhaps we migrate windows tests (and get rid of appveyor), and migrate the MacOS tests that we can migrate.
> 
> This will keep the number of distinct systems for CIs we use at 2 and should help reduce MacOS timeouts we experience on Travis.
> 
> @kalyanac -- What are your thoughts?

@nolanmar511 @kalyanac Having single config file to maintain would be nice. Maybe we can just pay for GitHub actions?

wyk9787 commented:
> > The Travis timeouts are pretty annoying.
> > Perhaps we migrate windows tests (and get rid of appveyor), and migrate the MacOS tests that we can migrate.
> > This will keep the number of distinct systems for CIs we use at 2 and should help reduce MacOS timeouts we experience on Travis.
> > @kalyanac -- What are your thoughts?
> 
> @nolanmar511 @kalyanac Having single config file to maintain would be nice. Maybe we can just pay for GitHub actions?

Github Actions are free for public repo: https://help.github.com/en/github/setting-up-and-managing-billing-and-payments-on-github/about-billing-for-github-actions

nolanmar511 commented:
Having a single config is nice; and I've been super happy with GitHub actions in cloud-profiler-nodejs and pprof-nodejs.

@wyk9787 -- Are we able to run all tests we run with GitHub actions that we currently run with Travis / Appveyor? If not, what limitations are there?

wyk9787 commented:
> Having a single config is nice; and I've been super happy with GitHub actions in cloud-profiler-nodejs and pprof-nodejs.
> 
> @wyk9787 -- Are we able to run all tests we run with GitHub actions that we currently run with Travis / Appveyor? If not, what limitations are there?

See "Current Limitations" in PR descriptions. In addition, running on Windows may be an issue if issue 2 in PR description can not be solved.

kalyanac commented:
Timeouts are very annoying and slows down the process. We can migrate to GitHub actions and only use Travis CI for configs not currently supported by GA. 

We should also take a look at what versions of OS X we want to use in these tests. The Travis CI uses pretty old versions.

kalyanac commented:
Re Windows issue, can you try running manually on Windows to see if there are any hidden characters that are failing the test. You can start a Windows VM on a GCP VM.

wyk9787 commented:
* Separated macOS and Linux testing. 
* Used specific version for OS and Xcode per Github Action policy.
* Changes to travis.ci will be introduced in next PR.

This PR is ready for review.

kalyanac commented:
I restarted Travis 2-3 times. It keeps timing out.

aalexand commented:
@kalyanac @wyk9787 Does the PR description accurately reflect the work done in this PR and the future plan?

wyk9787 commented:
> @kalyanac @wyk9787 Does the PR description accurately reflect the work done in this PR and the future plan?

Yes. I added one last sentence "...remove {Go, OS, XCode} sections that are covered by this Github action script." to better describe the next move.

wyk9787 commented:
@aalexand Is it okay to merge this PR?

aalexand commented:
> @aalexand Is it okay to merge this PR?

Sorry, this might be just me, but I can't easily grok the full plan from the PR description. It's long and contains many details, but what I'm looking for to understand is a summary of:

* What this PR does and why.
* What this PR doesn't do (i.e. the delta from the desired state), why and when that's going to be done.

Additional details are good, but the above is what I'm looking for.

Thanks for bearing with me :)

wyk9787 commented:
> > @aalexand Is it okay to merge this PR?
> 
> Sorry, this might be just me, but I can't easily grok the full plan from the PR description. It's long and contains many details, but what I'm looking for to understand is a summary of:
> 
> * What this PR does and why.
> * What this PR doesn't do (i.e. the delta from the desired state), why and when that's going to be done.
> 
> Additional details are good, but the above is what I'm looking for.
> 
> Thanks for bearing with me :)

Yes, I agree. It was definitely a bit messy.

Modified description to have sections and emphasis the "done" and "to-do".

aalexand commented:
> > > @aalexand Is it okay to merge this PR?
> > 
> > 
> > Sorry, this might be just me, but I can't easily grok the full plan from the PR description. It's long and contains many details, but what I'm looking for to understand is a summary of:
> > 
> > * What this PR does and why.
> > * What this PR doesn't do (i.e. the delta from the desired state), why and when that's going to be done.
> > 
> > Additional details are good, but the above is what I'm looking for.
> > Thanks for bearing with me :)
> 
> Yes, I agree. It was definitely a bit messy.
> 
> Modified description to have sections and emphasis the "done" and "to-do".

Thanks. Do I get it correctly that once we test it and it works, we should be able to at least get rid of the Appveyor config?

wyk9787 commented:
> > > > @aalexand Is it okay to merge this PR?
> > > 
> > > 
> > > Sorry, this might be just me, but I can't easily grok the full plan from the PR description. It's long and contains many details, but what I'm looking for to understand is a summary of:
> > > 
> > > * What this PR does and why.
> > > * What this PR doesn't do (i.e. the delta from the desired state), why and when that's going to be done.
> > > 
> > > Additional details are good, but the above is what I'm looking for.
> > > Thanks for bearing with me :)
> > 
> > 
> > Yes, I agree. It was definitely a bit messy.
> > Modified description to have sections and emphasis the "done" and "to-do".
> 
> Thanks. Do I get it correctly that once we test it and it works, we should be able to at least get rid of the Appveyor config?

I certainly hope so but I am not too confident. See issue #2 in the last section.

aalexand commented:
> > 
> > Thanks. Do I get it correctly that once we test it and it works, we should be able to at least get rid of the Appveyor config?
> 
> I certainly hope so but I am not too confident. See issue no. 2 in the last section.

I don't get the diff there - what's different? Visually these are the same, am I missed something? Or, line endings?..

wyk9787 commented:
> > > Thanks. Do I get it correctly that once we test it and it works, we should be able to at least get rid of the Appveyor config?
> > 
> > 
> > I certainly hope so but I am not too confident. See issue no. 2 in the last section.
> 
> I don't get the diff there - what's different? Visually these are the same, am I missed something? Or, line endings?..

I don't get the diff either :) 
But somehow the unit tests fail with this error message, showing me the diff that is the same.

aalexand commented:
> > > > Thanks. Do I get it correctly that once we test it and it works, we should be able to at least get rid of the Appveyor config?
> > > 
> > > 
> > > I certainly hope so but I am not too confident. See issue no. 2 in the last section.
> > 
> > 
> > I don't get the diff there - what's different? Visually these are the same, am I missed something? Or, line endings?..
> 
> I don't get the diff either :)
> But somehow the unit tests fail with this error message, showing me the diff that is the same.

So line endings?

wyk9787 commented:
> > > > > Thanks. Do I get it correctly that once we test it and it works, we should be able to at least get rid of the Appveyor config?
> > > > 
> > > > 
> > > > I certainly hope so but I am not too confident. See issue no. 2 in the last section.
> > > 
> > > 
> > > I don't get the diff there - what's different? Visually these are the same, am I missed something? Or, line endings?..
> > 
> > 
> > I don't get the diff either :)
> > But somehow the unit tests fail with this error message, showing me the diff that is the same.
> 
> So line endings?

FYI, @kalyanac and @nolanmar511 

You are right.
After learning Windows dev 101 (very fun indeed), I was finally able to reproduce the error on a Windows VM.
It turns out when using https://golang.org/pkg/io/ioutil/#ReadFile to read test file, it will add a "\r" in addition to "\n" at the end of the line on Windows. It makes sense since on Windows, "\r\n" together indicates a new line.

However, any pprof output (e.g. a dot file) does not include "\r", and that's why we are seeing the "weird" diff mismatch.

`ioutil.ReadFile` is widely used within pprof tests. Any immediate idea on how this issue can be easily addressed?

aalexand commented:
> `ioutil.ReadFile` is widely used within pprof tests. Any immediate idea on how this issue can be easily addressed?

`strings.ReplaceAll(s, "\r\n", "\n")` in the [Diff()](https://github.com/google/pprof/blob/03e1cf38a040dd596c700aa260c3eddc2bf38987/internal/proftest/proftest.go#L33) function used in the tests seems the most straightforward fix perhaps.

aalexand commented:
> It turns out when using https://golang.org/pkg/io/ioutil/#ReadFile to read test file, it will add a "\r" in addition to "\n" at the end of the line on Windows. It makes sense since on Windows, "\r\n" together indicates a new line.

I am not sure that ReadFile has that logic. I would expect that what happens is that when pprof prints a line it uses "\n" which on Windows expands to "\r\n". So the golden files read from disk have just "\n", but the reference, pprof-generated output has "\r\n". If ReadFile would do the expansion, the test would pass since both golden and reference files are read from disk.

So I think you can ignore my previous suggestion and just add the line endings replacement here - https://github.com/google/pprof/blob/03e1cf38a040dd596c700aa260c3eddc2bf38987/internal/driver/driver_test.go#L196.

aalexand commented:
Is it possible to add an icon / thumbnail on the main page to indicate the status of this new CI build, like we have for Travis and Appveyor?

wyk9787 commented:
> > It turns out when using https://golang.org/pkg/io/ioutil/#ReadFile to read test file, it will add a "\r" in addition to "\n" at the end of the line on Windows. It makes sense since on Windows, "\r\n" together indicates a new line.
> 
> I am not sure that ReadFile has that logic. I would expect that what happens is that when pprof prints a line it uses "\n" which on Windows expands to "\r\n". So the golden files read from disk have just "\n", but the reference, pprof-generated output has "\r\n". If ReadFile would do the expansion, the test would pass since both golden and reference files are read from disk.
> 
> So I think you can ignore my previous suggestion and just add the line endings replacement here -
> 
> https://github.com/google/pprof/blob/03e1cf38a040dd596c700aa260c3eddc2bf38987/internal/driver/driver_test.go#L196
> 
> .

I was actually looking at the failed tests in `internal/graph/...`. I was printing the ascii values from both `got` and `want` in https://github.com/google/pprof/blob/03e1cf38a040dd596c700aa260c3eddc2bf38987/internal/graph/dotgraph_test.go#L190 using this function:
```go
func printAscii(chunk []byte) {
  fmt.Println(“START”)
  for _, b := range chunk {
    fmt.Printf(“%d “, b)
  }
  fmt.Println(“DONE”)
}
```
It was actually `want` (golden files read from disk) which has 13 ("\r"'s ascii value) followed by a 10 ("\n"'s ascii value) in the output. `got` (output generated by pprof, and was not written to a file in this case) only has 10 in the out.

But you are right, for tests in `driver_test.go`, pprof generated outputs are first written to the disk and then read from disk using `ReadFile()`.  So something else is going on here.

> strings.ReplaceAll(s, "\r\n", "\n") in the Diff() function used in the tests seems the most straightforward fix perhaps.

This will not cut it here since `Diff()` is called after the bytes comparison: https://github.com/google/pprof/blob/03e1cf38a040dd596c700aa260c3eddc2bf38987/internal/graph/dotgraph_test.go#L197.

wyk9787 commented:
> Is it possible to add an icon / thumbnail on the main page to indicate the status of this new CI build, like we have for Travis and Appveyor?

Yes, see the third bullet point of "follow-up plan" in PR description.

Will add it in the next PR following https://help.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow#adding-a-workflow-status-badge-to-your-repository.

aalexand commented:
> It was actually `want` (golden files read from disk) which has 13 ("\r"'s ascii value) followed by a 10 ("\n"'s ascii value) in the output. `got` (output generated by pprof, and was not written to a file in this case) only has 10 in the out.

I guess you need to look at whether it's Git here that replaces the line endings on checkout then.

wyk9787 commented:
> > It was actually `want` (golden files read from disk) which has 13 ("\r"'s ascii value) followed by a 10 ("\n"'s ascii value) in the output. `got` (output generated by pprof, and was not written to a file in this case) only has 10 in the out.
> 
> I guess you need to look at whether it's Git here that replaces the line endings on checkout then.

I would have never thought it was actually Git that changes the line endings. Thanks!
Added a .gitattributes file to address this issue. Now tests are passing: https://github.com/wyk9787/pprof/actions/runs/147825124.

Also modified PR description.

aalexand commented:
@wyk9787 For the internal package import errors: in Travis config we use `go_import_path: github.com/google/pprof` setting to overcome that. Is there something similar in GH actions? If not, how are others dealing with this? It would be useful to figure this out, otherwise we won't know that the tests pass when someone sends us a PR?

aalexand commented:
> @wyk9787 For the internal package import errors: in Travis config we use `go_import_path: github.com/google/pprof` setting to overcome that. Is there something similar in GH actions? If not, how are others dealing with this? It would be useful to figure this out, otherwise we won't know that the tests pass when someone sends us a PR?

@wyk9787 https://laszlo.cloud/setting-gopath-in-github-actions#controlling-the-checkout-path ?

wyk9787 commented:
> > @wyk9787 For the internal package import errors: in Travis config we use `go_import_path: github.com/google/pprof` setting to overcome that. Is there something similar in GH actions? If not, how are others dealing with this? It would be useful to figure this out, otherwise we won't know that the tests pass when someone sends us a PR?
> 
> @wyk9787 https://laszlo.cloud/setting-gopath-in-github-actions#controlling-the-checkout-path ?

I was using a custom path for checkout but I included a variable `GITHUB_REPOSITORY`, which will lead to "wyk9787/pprof" for my testing branch. I thought the package checking is more than checking the path, so I didn't realize this was the issue.

Your comments reminded me to change it to hardcoded "google/pprof" in the path and it works now :)
https://github.com/wyk9787/pprof/runs/809235334?check_suite_focus=true

Thanks again for helping with both of those issues!
